File: prep/implementation/patterns_reference_implementation/modified_binary_search/MinimumDifferenceElement.java
import java.util.*;
import java.util.concurrent.*;
import java.util.function.*;

public class MinimumDifferenceElement {
  public static void main(String[] args) {
    try (Scanner in = new Scanner(System.in)) {
      int tt = in.nextInt();

      while (tt-- > 0) {
        int n = in.nextInt();
        int k = in.nextInt();

        int[] a = new int[n];
        for (int i = 0; i < n; i++) {
          a[i] = in.nextInt();
        }

        System.out.println(findClosestDiffElement(a, k));
      }
    }
  }

  // O(logn) / O(1)
  public static int findClosestDiffElement(int[] a, int k) {
    if (k < a[0]) {
      return a[0];
    }

    if (k > a[a.length - 1]) {
      return a[a.length - 1];
    }

    int low = 0, high = a.length - 1;
    while (low <= high) {
      int mid = low + (high - low) / 2;

      if (a[mid] < k) {
        low = mid + 1;
      } else if (a[mid] > k) {
        high = mid - 1;
      } else {
        return a[mid];
      }
    }

    return (Math.abs(k - a[low]) < Math.abs(k - a[high]) ? a[low] : a[high]);
  }
}

File: lc/coreprep/implementation/practice_iterations/iteration_archives/modified_binary_search_1/minimum_difference_element.cpp
#include <algorithm>
#include <iostream>
#include <vector>

using namespace std;

// O(logn) / O(1)
int minimum_difference(const vector<int> &a, int low, int high, int k) {
  if (k < a[low]) {
    return a[low];
  }

  if (k > a[high]) {
    return a[high];
  }

  while (low <= high) {
    int mid = low + (high - low) / 2;

    if (a[mid] == k) {
      return a[mid];
    } else if (a[mid] < k) {
      low = mid + 1;
    } else {
      high = mid - 1;
    }
  }

  return (abs(k - a[low]) < abs(k - a[high])) ? a[low] : a[high];
}

int main() {
  ios::sync_with_stdio(0);
  cin.tie(0);

  int tt, n, k;
  cin >> tt;

  while (tt--) {
    cin >> n >> k;

    vector<int> a(n);
    for (int i = 0; i < n; i++) {
      cin >> a[i];
    }

    cout << minimum_difference(a, 0, n - 1, k) << "\n";
  }

  return 0;
}